# Quantum Teleportation

Quantum teleportation is a protocol used to transfer the quantum state of a qubit from one location to another.

The interesting part is that the qubit itself is not physically transported.

Instead, the protocol uses:
- entanglement
- quantum measurement
- classical communication

to reconstruct the original quantum state somewhere else.

Quantum teleportation does not teleport matter or energy.

What gets transferred is the quantum state information.

The original state gets destroyed during the process.

The protocol involves three qubits:

- qubit 0:
  the unknown state Alice wants to send

- qubit 1:
  Alice’s half of an entangled Bell pair

- qubit 2:
  Bob’s half of the Bell pair

Initially:
- Alice and Bob share entanglement
- Alice performs operations and measurements
- Bob later applies corrections using Alice’s classical measurement results

# Why Entanglement Matters

Without entanglement, the quantum state cannot be reconstructed remotely.

The Bell pair acts like a shared quantum resource between Alice and Bob.

The protocol works because operations performed on entangled systems affect the combined quantum state rather than isolated qubits independently.

At first teleportation can feel impossible because it seems like information appears at Bob’s side instantly.

But Bob cannot reconstruct the state until Alice sends her classical measurement bits.

So the protocol still respects the speed limit imposed by relativity.

The teleportation happens through a combination of:
- shared entanglement
- classical communication
- quantum state collapse

# No-Cloning Principle

A fundamental rule in quantum mechanics is that an unknown quantum state cannot be copied perfectly.

This is very different from classical information, where files or data can be duplicated freely.

Quantum teleportation does not violate this principle because the original state gets destroyed during measurement while the state is reconstructed on Bob’s side.

# Why Teleportation Does Not Break Relativity

At first quantum teleportation can seem like faster-than-light communication because entanglement creates instant correlations between qubits.

However, Bob still cannot reconstruct the state until Alice sends her classical measurement bits through ordinary communication.

Since classical information cannot travel faster than light, the protocol still obeys relativity.

# State We Will Teleport

For simplicity, the unknown state will begin as:

|1>

Later, more general superposition states can also be teleported using the same protocol.

In [1]:
from qiskit import QuantumCircuit
from qiskit_aer import AerSimulator

# create circuit w/ 3 qubits and 2 classical bits
qc = QuantumCircuit(3,2)

# step 1: prepare state to teleport

# prepare |1>
qc.x(0)

# step 2: create bell pair between qubit 1 and qubit 2

qc.h(1)
qc.cx(1,2)

# step 3: alice entangles her qubits

qc.cx(0,1)
qc.h(0)

# step 4: alice measures her qubits

qc.measure(0,0)
qc.measure(1,1)

# step 5:bob applies corrections based on alice's measurement

qc.cx(1,2)
qc.cz(0,2)

# measure bob's qubit
qc.measure_all()

# draw circuit
print(qc.draw())

# run 
simulator = AerSimulator()

job = simulator.run(qc, shots=1000)

result = job.result()
counts = result.get_counts()

print(counts)

        ┌───┐          ┌───┐┌─┐         ░ ┌─┐      
   q_0: ┤ X ├───────■──┤ H ├┤M├──────■──░─┤M├──────
        ├───┤     ┌─┴─┐└┬─┬┘└╥┘      │  ░ └╥┘┌─┐   
   q_1: ┤ H ├──■──┤ X ├─┤M├──╫───■───┼──░──╫─┤M├───
        └───┘┌─┴─┐└───┘ └╥┘  ║ ┌─┴─┐ │  ░  ║ └╥┘┌─┐
   q_2: ─────┤ X ├───────╫───╫─┤ X ├─■──░──╫──╫─┤M├
             └───┘       ║   ║ └───┘    ░  ║  ║ └╥┘
   c: 2/═════════════════╩═══╩═════════════╬══╬══╬═
                         1   0             ║  ║  ║ 
meas: 3/═══════════════════════════════════╩══╩══╩═
                                           0  1  2 
{'111 11': 255, '100 00': 253, '101 01': 248, '110 10': 244}


# Understanding the Output

The original state being teleported was:

|1>

The measurement results looked like:

111 11
100 00
101 01
110 10

Initially this looked confusing because multiple outputs appeared instead of one fixed result.

The important thing to notice is that Alice’s measurement results are random.

This is expected because quantum measurement causes probabilistic collapse.

Another important detail is that the first bit in the left group corresponds to Bob’s qubit.

For example:

111
100
101
110

all begin with:

1

This shows that Bob’s final qubit consistently reconstructed the original state:

|1>

even though Alice’s own measurement bits varied randomly.

The teleportation protocol works because:
- entanglement creates shared quantum correlations
- Alice’s measurement transfers information into the entangled system
- classical bits tell Bob which correction to apply
- Bob reconstructs the original quantum state

So the state itself was never copied or directly sent.

Instead, the quantum information was transferred using entanglement and classical communication.